### PR TITLE
Add additional fallback to http.publish_port and restrict fallback to transport.publish_port

### PR DESCRIPTION
**Pull request by ywelsch**
PR #14535 fixed an issue where the port of a publish address would possibly not match the port of the corresponding bound address. This was implemented by matching publish to bound addresses and selecting the port of the bound address that matched. In case of http, we would fail if we did not find a match. For tcp transport, we would (similar to old behavior) not fail but just chose the port of the first bound address.

This behavior makes it difficult to properly set up network settings in case where publish address does not match any of the bound addresses, for example in case of Docker containers (see https://discuss.elastic.co/t/upgrading-to-es-2-2-0-doesnt-allow-node-to-have-different-bind-and-publish-addresses/41290 ).

I suggest the following change in fallback behavior:

In case where the publish address cannot be matched to any of the bound addresses and an explicit publish_port setting is not provided, the publish_port is now selected as the unique port of bound addresses. An exception is thrown in case of ambiguities. This applies to both tcp and http. The fallback now ensures that when the user has specified a concrete value for http.port (not a range), http.bind_port also falls back to that one. In case of transport.publish_port it is a bit more restrictive: Instead of chosing the port of the first bound address, we require that the port of bound addresses is unique.

### CODE REVIEW

**Comment by ywelsch** [comment]
@rmuir can you have a look at this one?

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java [hunk 1]
```diff
@@ -343,12 +344,23 @@ protected void doStart() {
...
+            for (InetSocketTransportAddress boundAddress : boundAddresses) {
+                ports.add(boundAddress.getPort());
+            }
+            if (ports.size() == 1) {
+                publishPort = ports.iterator().next().value;
+            }
+        }
+
+        if (publishPort < 0) {
+            throw new BindHttpException(SETTING_HTTP_PUBLISH_PORT.getKey() + " is ambiguous (bound addresses: " + boundAddresses + "). " +
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java]: SETTING_HTTP_PUBLISH_PORT is likely empty right? In that case I think this message will be confusing. Maybe something like "can't auto-resolve http publish port, multiple bounded ports and none of them matched the publish host. Please specify a unique port by setting" + SETTING_HTTP_PORT.getKey() + " or " + SETTING_HTTP_PUBLISH_PORT.getKey()

File: core/src/main/java/org/elasticsearch/transport/netty/NettyTransport.java [hunk 2]
```diff
@@ -553,13 +555,20 @@ private BoundTransportAddress createBoundTransportAddress(String name, Settings
...
+            for (InetSocketAddress boundAddress : boundAddresses) {
+                ports.add(boundAddress.getPort());
+            }
+            if (ports.size() == 1) {
+                publishPort = ports.iterator().next().value;
+            }
+        }
+
+        if (publishPort < 0) {
+            throw new BindTransportException("publish_port is ambiguous (bound addresses: " + boundAddresses + "). " +
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/netty/NettyTransport.java]: same comment about messaging.

File: core/src/test/java/org/elasticsearch/http/netty/HttpPublishPortIT.java [hunk 3]
```diff
@@ -34,18 +35,37 @@
...
-    protected Settings nodeSettings(int nodeOrdinal) {
-        return Settings.settingsBuilder()
-                .put(super.nodeSettings(nodeOrdinal))
-                .put(NetworkModule.HTTP_ENABLED.getKey(), true)
-                .put(HttpTransportSettings.SETTING_HTTP_PUBLISH_PORT.getKey(), 9080)
-                .build();
-    }
 
     public void testHttpPublishPort() throws Exception {
+        Settings settings = Settings.settingsBuilder()
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/http/netty/HttpPublishPortIT.java]: can we use package private methods and have unit tests for this.. an integration seems like an overkill.

**Comment by rmuir** [comment]
I honestly think this is getting too complex. We should really open an issue to remove this publish host alltogether. 

If node B is talking to node A, instead of allowing node A to "lie about his phone number", node B should just look at the caller ID and not trust him at all. We are talking about TCP here, this is infallible.

This means just use https://docs.oracle.com/javase/7/docs/api/java/net/Socket.html#getRemoteSocketAddress%28%29

That's how every other networking app works and why we see so much confusion around this publish host.

**Comment by bleskes** [comment]
I think the change makes sense. Left some comments. I would also like to have this clearly documented some where...

**Comment by bleskes** [comment]
I share @rmuir sentiment that this is getting out of hand and agree we should re-evaluate long term. I don't think it should stop this fix though...

**Comment by ywelsch** [comment]
Rebased on current master and added two commits to address the comments. Can you have another look @bleskes ?

**Comment by bleskes** [comment]
The change LGTM. Do we have a test for the NettyTransport case? (i.e. not http)

**Comment by ywelsch** [comment]
@bleskes I've added one just now. Also randomized the port in the test.

**Comment by bleskes** [comment]
Thanks @ywelsch. I do think we need to beef up testing. We need at least:

1) Multiple bound addresses
2) Sometimes they should all have different ports (and we resolve publish address to one of them and succeed or check that we fail o.w.)
3) Sometimes they should all have the same port and we should succeed both in the case we specify a publish port or not.
4) Make sure that when we do specific a publish port it is always used.

IMO these test don't even need to start the transport, I hope we can make the relevant methods static and pass what they need to them.

**Comment by bleskes** [comment]
LGTM. Thanks for the tests!

**Pull request merged**